core/store/src/trie/iterator.rs

```diff
@@ -197,22 +206,33 @@ impl<'a> TrieIterator<'a> {
         prefix
     }
 
-    /// Returns hashes of nodes with paths in [path_begin, path_end). Used by state parts
+    fn eval_predicate(&self, predicate: Option<&dyn Fn(&[u8]) -> bool>) -> bool {
+        match predicate {
+            Some(f) => f(&self.key()),
+            None => false,
+        }
+    }
+
+    /// Visits all nodes belonging to the interval [path_begin, path_end) in depth-first search order.
+    /// Returns vector of pairs with node hash and evaluation of predicate on node key for all nodes found.
+    /// Used to generate and apply state parts for state sync.
     pub(crate) fn visit_nodes_interval(
         &mut self,
         path_begin: &[u8],
         path_end: &[u8],
-    ) -> Result<Vec<CryptoHash>, StorageError> {
+        predicate: Option<&dyn Fn(&[u8]) -> bool>,
+    ) -> Result<Vec<(CryptoHash, bool)>, StorageError> {
         let path_begin_encoded = NibbleSlice::encode_nibbles(path_begin, true);
         let last_hash = self.seek_nibble_slice(NibbleSlice::from_encoded(&path_begin_encoded).0)?;
         let mut prefix = Self::common_prefix(path_end, &self.key_nibbles);
         if self.key_nibbles[prefix..] >= path_end[prefix..] {
             return Ok(vec![]);
         }
         let mut nodes_list = Vec::new();
+
         // Actually (self.key_nibbles[..] == path_begin) always because path_begin always ends in a node
         if &self.key_nibbles[..] >= path_begin {
-            nodes_list.push(last_hash);
+            nodes_list.push((last_hash, self.has_value() && self.eval_predicate(predicate)));
```

Comment:
It is still not no-op when there is no predicate. Also I still don't like this API. @matklad I wonder whether you have ideas regarding how this could be done better.

Reply:
:thinking: yeah, I am having hard time wrapping my head around the API here. It feels suspicious that we seem to do `retrieve_raw_bytes(&hash)?;` twice, for example. We also return internal node hashes, rather than proper key-values.

Don't have immediate suggestions here, besides:

* return a tripple of `(NodeHash, Value, Key)`, to let the caller figure how to deal with them
* rather than returning a vec, maybe cast the thing as a proper iterator:

```
struct TrieIntervalIterator {
    inner: TrieIterator,
    interval: std::ops::Range<&[u8]>
}
```

Reply:
Hm, one think that feels a bit strange is that we tend to allocate all over the place, when reading both keys and values. I wonder if this can benefit from *not* using the iterator API, and instead using "streaming" iterators, which borrow from the current iterator object. Something like

```rust
struct StreamingTrieIterator {
    trie: &'a Trie,
    trail: Vec<Crumb>,
    pub(crate) key_nibbles: Vec<u8>,
    root: CryptoHash,
    
    key_buf: Vec<u8>,
    val_buf: Vec<u8>,
}

struct Item<'a> {
    pub key: &'a [u8],
    pub value: &'a [u8],
}

impl StreamingTrieIterator {
    pub fn next<'a>(&mut self) -> Item<'a> // Item borrows *current* key and value from `buf` fields
}
```

It seems that this can make access to the current key cheap, which means that we'll be able to just always return the full key alongside the value, which means the caller will be able to calculate arbitrary predicates using always available key.

Reply:
@Longarithm do we actually need the hash if the predicate returns false?

Reply:
@bowenwang1996 We need to return all hashes from this method. They are later converted to `TrieChanges` and nodes mentioned there are applied to `ShardTries`.

Reply:
After discussion with @matklad I moved logic of computing predicate out of there and started returning pairs (hash, key) from it, which makes `visit_nodes_interval` more clear

The average overhead of this change in `visit_nodes_interval` is not more than 1%, which was computed by running 500 iterations of `test_combine_state_parts` test.